bitcoin/src/blockdata/script/owned.rs

```diff
@@ -179,17 +179,37 @@ impl ScriptBuf {
     /// for a P2WPKH output. The `scriptCode` is described in [BIP143].
     ///
     /// [BIP143]: <https://github.com/bitcoin/bips/blob/99701f68a88ce33b2d0838eb84e115cef505b4c2/bip-0143.mediawiki>
+    #[deprecated(note = "use `segwit_script_code` instead")]
     pub fn p2wpkh_script_code(&self) -> Option<ScriptBuf> {
-        self.v0_p2wpkh().map(|wpkh| {
-            Builder::new()
+        // Unlike `segwit_script_code`, this method returns `None` for anything
+        // non-v0_p2wpkh.
+        if self.is_v0_p2wpkh() {
+            Some(self.segwit_script_code().into_script_buf())
+        } else {
+            None
+        }
+    }
+
+    /// Converts a script into a Segwit spending script code. Respectively,
+    /// P2WPKH is handled in a special manner as defined by [BIP143] while every
+    /// other script is left untouched.
+    ///
+    /// [BIP143]: <https://github.com/bitcoin/bips/blob/99701f68a88ce33b2d0838eb84e115cef505b4c2/bip-0143.mediawiki>
```

Comment:
In 4cc399b0844be1198fe060f9ca5e05f728735be2:

Should mention in the comments that if you use this with a non-segwit script, you'll get an invalid script code. I think this is fine, it'd be more of a PITA to return an `Option` and make people unwrap, but we need to document it.

Reply:
Hmmm, actually, given that this is the only way to construct a `SegwitScriptCode` maybe we *should* enforce that we're starting from a segwit script. Then users know that if they have a `SegwitScriptCode` that it's actually a segwit script code.

As written, you can use this method to turn arbitrary sequences of bytes into a "segwit script code", which feels a little footgunny.

Reply:
Fair enough. Do you have a design pattern in mind? It should be clear that `segwit_script_code_signature_hash` wants a valid `SegwitScriptCode` for P2WPKH, and everything else can be handled as-is.

Reply:
Maybe we convert `SegwitScriptCode` into an enum (and call it something else), then that enum has two variants:

```rust
enum X {
    SegwitScriptCode(...),
    Arbitrary(...)
}
```

and then `X` makes an explicit check for each when converting from `ScriptBuf`.

Reply:
I think the existing API is probably fine for now.

We're stepping into a rat's nest here ... I think an "ideal" API would somehow have a `ScriptPubKey` type which was maybe an enum of all the different scriptpubkey types, and then we'd have one `script_code` method that looked at the enum and then returned an enum of different scriptcode types... or something. I don't know. Every time I try to think about this the API complexity goes out of control and users wind up needing

Reply:
So with that said, I think the two sensible APIs here are:

* what you've done
* an alternate which returns an `Option` and returns `None` if you pass it a non-segwit script.

I'd be fine with either but have a slight preference for the first.